programs/st_client.c

```diff
@@ -0,0 +1,354 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the project nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.	IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifdef _WIN32
+#define _CRT_SECURE_NO_WARNINGS
+#endif
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdarg.h>
+#include <sys/types.h>
+#ifndef _WIN32
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <errno.h>
+#include <pthread.h>
+#include <unistd.h>
+#else
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#endif
+#include <usrsctp.h>
+#include "programs_helper.h"
+
+#define MAX_PACKET_SIZE (1<<16)
+#define BUFFER_SIZE 80
+#define DISCARD_PPID 39
+
+#define TIMER_INTERVAL_MSECS 10
+
+static int connecting = 0;
+static int finish = 0;
+
+static unsigned get_tick_count()
+{
+#ifdef _WIN32
+	return GetTickCount();
+#else
+	static const clock_t clocks_per_msec = CLOCKS_PER_SEC / 1000;
+	return clock() / clocks_per_msec;
+#endif
+}
+
+static void
+handle_packets(int sock, struct socket* s, void* sconn_addr)
+{
+	char *dump_buf;
+	ssize_t length;
+	char buf[MAX_PACKET_SIZE];
+
+	fd_set rfds;
+	struct timeval tv;
+
+	unsigned next_fire_time = get_tick_count();
+	unsigned last_fire_time = next_fire_time;
+
+	while (!finish) {
+
+		unsigned now = get_tick_count();
+		int wait_time;
+
+		if ((int) (now - next_fire_time) > 0) {
+			usrsctp_handle_timers(now - last_fire_time);
+			last_fire_time = now;
+			next_fire_time = now + TIMER_INTERVAL_MSECS;
+		}
+
+		wait_time = next_fire_time - now;
+		tv.tv_sec = wait_time / 1000;
+		tv.tv_usec = (wait_time % 1000) * 1000;
+
+		select(1, &rfds, NULL, NULL, &tv);
+
+		length = recv(sock, buf, MAX_PACKET_SIZE, 0);
```

Comment:
I think you only want to do this if `FD_ISSET(sock, rdfs)` is true. If that is not true, you can't read anything from the socket.

Reply:
Hi @tuexen. Honestly this `st_client.c` is at it was in the original PR. Indeed we cannot make it work. I'm adding changes told by you and also documenting into it how to use it:

```c
/*
 * Usage: st_client local_addr local_port remote_addr remote_port remote_sctp_port
 */
```

I've added some prints into `st_client.c` to figure out what is happening:

```c
		wait_time = next_fire_time - now;
		tv.tv_sec = wait_time / 1000;
		tv.tv_usec = (wait_time % 1000) * 1000;

		FD_ZERO(&rfds);
		FD_SET(sock, &rfds);

		if (FD_ISSET(sock, &rfds))
			printf("--- FD_ISSET(sock, &rfds): yes\n");

		printf("--- calling select()\n");

		select(sock + 1, &rfds, NULL, NULL, &tv);

		printf("--- calling recv() 2\n");

		length = recv(sock, buf, MAX_PACKET_SIZE, 0);

		printf("--- recv() returned, length:%zu\n", length);

		if (length > 0) {
			if ((dump_buf = usrsctp_dumppacket(buf, (size_t)length, SCTP_DUMP_INBOUND)) != NULL) {
				fprintf(stderr, "%s", dump_buf);
				usrsctp_freedumpbuffer(dump_buf);
			}
			usrsctp_conninput(sconn_addr, buf, (size_t)length, 0);
		}
	}
````

Testing it:

```bash
# run server
$ ./echo_server 11111 22222

# run client
$ ./st_client 127.0.0.1 22222 127.0.0.1 11111 11111
```

It produces zero output in the server and this output in the client:

```
O 16:20:33.113793 0000 f9 7d 2b 67 00 00 00 00 68 43 df cd 01 00 00 56 1c ce f2 40 00 02 00 00 00 0a 08 00 08 eb c4 6b c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 f6 82 90 8c a6 0d 41 8d ef 6c 26 22 c5 43 3c ec 43 25 62 64 28 8e 77 42 de ab 4a e8 92 45 99 aa 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET
--- FD_ISSET(sock, &rfds): yes
--- calling select()
--- calling recv() 2
```

and it gets stuck in the `length = recv(sock, buf, MAX_PACKET_SIZE, 0);` line.

Reply:
Try something like:
```
		wait_time = next_fire_time - now;
		tv.tv_sec = wait_time / 1000;
		tv.tv_usec = (wait_time % 1000) * 1000;

		FD_ZERO(&rfds);
		FD_SET(sock, &rfds);

		printf("--- calling select()\n");

		select(sock + 1, &rfds, NULL, NULL, &tv);
		if (FD_ISSET(sock, &rfds)) {
			printf("--- FD_ISSET(sock, &rfds): yes\n");

			printf("--- calling recv() 2\n");

			length = recv(sock, buf, MAX_PACKET_SIZE, 0);

			printf("--- recv() returned, length:%zu\n", length);

			if (length > 0) {
				if ((dump_buf = usrsctp_dumppacket(buf, (size_t)length, SCTP_DUMP_INBOUND)) != NULL) {
					fprintf(stderr, "%s", dump_buf);
					usrsctp_freedumpbuffer(dump_buf);
				}
				usrsctp_conninput(sconn_addr, buf, (size_t)length, 0);
			}
		} else {
			printf("--- FD_ISSET(sock, &rfds): no\n");
		}
	}
```

Reply:
I'm afraid it just produces this output in the client (and nothing in the server):

```
$ ./st_client 127.0.0.1 22222 127.0.0.1 11111 11111

O 16:42:57.634535 0000 fd 80 2b 67 00 00 00 00 7b 61 7a 09 01 00 00 56 74 5a ce b6 00 02 00 00 00 0a 08 00 ba 43 c8 36 c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 ca d1 11 0b cd 3b 90 b9 97 05 7c 28 e7 c9 b2 14 cd e1 91 be 2d 72 82 45 68 9c 1c e1 f1 17 db 56 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET
--- calling select()
--- FD_ISSET(sock, &rfds): no
--- calling select()
--- FD_ISSET(sock, &rfds): no
--- calling select()
--- FD_ISSET(sock, &rfds): no
--- calling select()
--- FD_ISSET(sock, &rfds): no
// etc etc etc
```

Reply:
I understand that, even if `st_client` command line arguments are a bit pain, I'm using them correctly (OSX):

```bash
$ lsof -i -n -P | grep -E '(echo_se|st_client)'

echo_serv 25145            ibc    4u  IPv4 0xb020ae295166cc0d      0t0  UDP *:11111
echo_serv 25145            ibc    5u  IPv6 0xb020ae2951670a95      0t0  UDP *:11111
st_client 25554            ibc    3u  IPv4 0xb020ae294236373d      0t0  UDP 127.0.0.1:22222->127.0.0.1:11111
```

Reply:
No sure why the server is listed twice... If you capture the traffic on the UDP interface, do you see an INIT-ACK encapsulated in UDP? If yes, then there is a problem on the client side. If no, there is a problem on the server side.
Another point: After 1 second there should be a transmission of the INIT. So a timer should run off. Does this happen?

Reply:
Somehow the echo server is not responding to the `st_client`:

```
16:52:30.607280 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
16:52:30.607293 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
16:52:33.610211 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
16:52:33.610226 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
```

However, it does reply to the `client` program:

```
16:53:49.217415 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 156
16:53:49.217459 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 156
16:53:49.217821 IP 127.0.0.1.11111 > 127.0.0.1.22222: UDP, length 564
16:53:49.217847 IP 127.0.0.1.11111 > 127.0.0.1.22222: UDP, length 564
16:53:49.218336 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 428
16:53:49.218366 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 428
16:53:49.221919 IP 127.0.0.1.11111 > 127.0.0.1.22222: UDP, length 16
16:53:49.221963 IP 127.0.0.1.11111 > 127.0.0.1.22222: UDP, length 16
16:53:49.222080 IP 192.168.1.55.11111 > 192.168.1.55.22222: UDP, length 56
16:53:49.222092 IP 192.168.1.55.11111 > 192.168.1.55.22222: UDP, length 56
16:53:49.222131 IP 5.5.5.5.11111 > 5.5.5.5.22222: UDP, length 56
16:53:49.222142 IP 5.5.5.5.11111 > 5.5.5.5.22222: UDP, length 56
16:53:49.222214 IP6 ::1.11111 > ::1.22222: UDP, length 56
16:53:49.222226 IP6 ::1.11111 > ::1.22222: UDP, length 56
16:53:49.223098 IP 192.168.1.55.22222 > 192.168.1.55.11111: UDP, length 56
16:53:49.223118 IP 192.168.1.55.22222 > 192.168.1.55.11111: UDP, length 56
16:53:49.223258 IP 5.5.5.5.22222 > 5.5.5.5.11111: UDP, length 56
16:53:49.223271 IP 5.5.5.5.22222 > 5.5.5.5.11111: UDP, length 56
16:53:49.223362 IP6 ::1.22222 > ::1.11111: UDP, length 56
16:53:49.223375 IP6 ::1.22222 > ::1.11111: UDP, length 56
16:53:49.224581 IP 192.168.1.55.11111 > 192.168.1.55.22222: UDP, length 56
16:53:49.224594 IP 192.168.1.55.11111 > 192.168.1.55.22222: UDP, length 56
16:53:49.224637 IP6 ::1.22222 > ::1.11111: UDP, length 56
16:53:49.224647 IP6 ::1.22222 > ::1.11111: UDP, length 56
16:53:49.225255 IP6 ::1.11111 > ::1.22222: UDP, length 56
16:53:49.225268 IP6 ::1.11111 > ::1.22222: UDP, length 56
16:53:49.225361 IP 192.168.1.55.22222 > 192.168.1.55.11111: UDP, length 56
16:53:49.225372 IP 192.168.1.55.22222 > 192.168.1.55.11111: UDP, length 56
16:53:49.225438 IP 5.5.5.5.22222 > 5.5.5.5.11111: UDP, length 56
16:53:49.225448 IP 5.5.5.5.22222 > 5.5.5.5.11111: UDP, length 56
16:53:49.225540 IP 5.5.5.5.11111 > 5.5.5.5.22222: UDP, length 56
16:53:49.225553 IP 5.5.5.5.11111 > 5.5.5.5.22222: UDP, length 56
```

Reply:
> No sure why the server is listed twice...

Because it's listening in IPv4 and IPv6.

> If you capture the traffic on the UDP interface, do you see an INIT-ACK encapsulated in UDP? If yes, then there is a problem on the client side. If no, there is a problem on the server side.
> Another point: After 1 second there should be a transmission of the INIT. So a timer should run off. Does this happen?

I just see traffic from `st_client` to `echo_server`, and yes there are retransmissions (I assume it's the INIT):

```
// run server:
$ ./echo_server 11111 22222

// run client:
$ ./st_client 127.0.0.1 22222 127.0.0.1 11111 11111

// stdout in client:

16:57:50.374429 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
16:57:50.374444 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100

// after 2 seconds:
16:57:53.380791 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
16:57:53.380814 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100

// after 2 seconds:
16:57:59.389498 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
16:57:59.389521 IP 127.0.0.1.22222 > 127.0.0.1.11111: UDP, length 100
```

Not sure if relevant, but the UDP packet size in the first message (SCTP INIT) in `client` is 156 bytes, while in `st_client` it's 100 bytes.

Reply:
I'm starting to understand that both `echo_server` and `client` use:

```
sock = usrsctp_socket(AF_INET6, SOCK_STREAM, IPPROTO_SCTP...
```

while in `st_client` it is:

```
usrsctp_socket(AF_CONN, SOCK_STREAM, IPPROTO_SCTP...
```

so obviously this won't work...

Reply:
and here we go:

* In the server:

```bash
$ ./ekr_server 127.0.0.1 11111 127.0.0.1 22222
```

* In the client:

```bash
$ ./st_client 127.0.0.1 22222 127.0.0.1 11111 11111
```

* Logs in the client:

```
O 17:20:18.738085 0000 e6 03 2b 67 00 00 00 00 db 5e ee 70 01 00 00 56 9f 90 25 4c 00 02 00 00 00 0a 08 00 5a 3b ef b4 c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 1c 50 e9 7e b2 02 5e 11 88 a4 2c 1d e8 f7 f8 63 78 2a e8 1b 29 1a 05 61 ab e7 4e d8 d0 49 92 84 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET
--- calling select()
--- FD_ISSET(sock, &rfds): yes
--- calling recv() 2
--- recv() returned, length:16
```

* Logs in the server:

```
I 17:20:18.739651 0000 e6 03 2b 67 00 00 00 00 db 5e ee 70 01 00 00 56 9f 90 25 4c 00 02 00 00 00 0a 08 00 5a 3b ef b4 c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 1c 50 e9 7e b2 02 5e 11 88 a4 2c 1d e8 f7 f8 63 78 2a e8 1b 29 1a 05 61 ab e7 4e d8 d0 49 92 84 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET

O 17:20:18.746047 0000 2b 67 e6 03 9f 90 25 4c 7a a4 17 76 06 00 00 04 # SCTP_PACKET
```

Obviously the `ekr_server` is not a echo server.

Reply:
Also, `st_client` can talk to itself:

```
$ ./st_client 127.0.0.1 11111 127.0.0.1 22222 11111
```

and

```
$ ./st_client 127.0.0.1 22222 127.0.0.1 11111 11111
```

Logs in the first st_client:

```
O 17:23:57.090391 0000 e8 07 2b 67 00 00 00 00 51 9d 2d f4 01 00 00 56 9d bf 3d d0 00 02 00 00 00 0a 08 00 ae 6b 52 c5 c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 98 76 2e 70 04 b2 19 55 21 41 ac 43 5e 57 df 97 b4 08 ae f1 1c 30 5d 2d 26 c2 c7 1f 85 d6 67 41 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET

I 17:23:57.091196 0000 2b 67 e8 07 9d bf 3d d0 f1 fb 6f 02 06 00 00 04 # SCTP_PACKET
```

Logs in the second st_client:

```
O 17:23:54.605069 0000 e8 e6 2b 67 00 00 00 00 e4 c2 d9 13 01 00 00 56 c4 05 38 55 00 02 00 00 00 0a 08 00 ba 1f f2 42 c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 44 d4 42 93 b6 ab 94 b1 05 da 28 e9 9b 47 5d 51 c9 70 e8 a7 1b 3e 47 36 85 83 13 48 e1 23 95 ea 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET

I 17:23:57.091071 0000 e8 07 2b 67 00 00 00 00 51 9d 2d f4 01 00 00 56 9d bf 3d d0 00 02 00 00 00 0a 08 00 ae 6b 52 c5 c0 00 00 04 80 08 00 09 c0 0f c1 80 82 00 00 00 80 02 00 24 98 76 2e 70 04 b2 19 55 21 41 ac 43 5e 57 df 97 b4 08 ae f1 1c 30 5d 2d 26 c2 c7 1f 85 d6 67 41 80 04 00 06 00 01 00 00 80 03 00 06 80 c1 00 00 # SCTP_PACKET
```